test/smart-wallet/signatures/EIP1271Test.ts

```diff
@@ -0,0 +1,589 @@
+          
+    let value = await smartAccountWithHandlerInterface.callStatic["isValidSignature(bytes32,bytes)"](dataHash, "0x");
+    let bytes4zero = "0x00000000";
+    expect(value).to.be.equal(bytes4zero);
+  });
+
+  it("Fallback handler returns bytes4(0) if signature is not valid", async function () { 
+    const dataHash = ethers.utils.keccak256("0xdeafbeef");
+    const invalidSignature = "0xabcdefdecafcafe0";
+    
+    let smartAccountWithHandlerInterface = await ethers.getContractAt(
+        "contracts/smart-contract-wallet/handler/DefaultCallbackHandler.sol:DefaultCallbackHandler",
+        signerSmartAccount.address
+    );
+          
+    let value = await smartAccountWithHandlerInterface.callStatic["isValidSignature(bytes32,bytes)"](dataHash, invalidSignature);
+    let bytes4zero = "0x00000000";
+    expect(value).to.be.equal(bytes4zero);
+  });
+
+  
+  it("Fallback handler returns magic value if message has been signed", async function () { 
+    const delegateCall = 1;
+    const dataToSign = "0xdeafbeefdecaf0";
+
+    // prepare tx to call signMessageLib.signMessage from signerSmartAccount 
+    // through delegate call 
+    const safeTx: SafeTransaction = buildSafeTransaction({
+      to: signMessageLib.address,
+      data: encodeSignMessage(dataToSign),
+      operation: delegateCall,
+      nonce: await signerSmartAccount.getNonce(EOA_CONTROLLED_FLOW),
+    });
+
+    const transaction: Transaction = {
+      to: safeTx.to,
+      value: safeTx.value,
+      data: safeTx.data,
+      operation: safeTx.operation,
+      targetTxGas: safeTx.targetTxGas,
+    };
+    const refundInfo: FeeRefund = {
+      baseGas: safeTx.baseGas,
+      gasPrice: safeTx.gasPrice,
+      tokenGasPriceFactor: safeTx.tokenGasPriceFactor,
+      gasToken: safeTx.gasToken,
+      refundReceiver: safeTx.refundReceiver,
+    };
+
+    const chainId = await mainSmartAccount.getChainId();
+
+    const { signer, data } = await safeSignTypedData(
+      accounts[0],
+      signerSmartAccount,
+      safeTx,
+      chainId
+    );
+
+    let signature = "0x";
+    signature += data.slice(2);
+
+    const txSignMessage = 
+      await signerSmartAccount.connect(accounts[0]).execTransaction(transaction, refundInfo, signature);
+    const receipt = await txSignMessage.wait();
+
+    let smartAccountWithHandlerInterface = await ethers.getContractAt(
+      "contracts/smart-contract-wallet/handler/DefaultCallbackHandler.sol:DefaultCallbackHandler",
+      signerSmartAccount.address
+    );
+
+    let dataHash = await smartAccountWithHandlerInterface.callStatic["getMessageHash(bytes)"](dataToSign);
+    
+    let eip1271MagicValue = "0x1626ba7e"; 
+    let value = await smartAccountWithHandlerInterface.callStatic["isValidSignature(bytes32,bytes)"](dataHash, "0x");
+    expect(value).to.be.equal(eip1271MagicValue);
+  
+  });
+
+  it("Fallback handler returns magic value if correct signature has been provided directly to Smart Account", async function () { 
```

Comment:
didn't get this case and how it works?
this is exactly same as ambire x opensea ux we checked right?

Reply:
yep. it's the case, when some off-chain script calls isValidSignature, not other contract

Reply:
interesting